pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go

```diff
@@ -166,96 +189,87 @@ func syncVolume(volumeIndex *persistentVolumeOrderedIndex, binderClient binderCl
 			// to start the volume again at the beginning of this lifecycle.
 			// ClaimRef is the last bind between persistent volume and claim.
 			// The claim has already been deleted by the user at this point
-			oldClaimRef := volume.Spec.ClaimRef
-			volume.Spec.ClaimRef = nil
-			_, err = binderClient.UpdatePersistentVolume(volume)
+			newVolume, err := copyVolume(volume)
 			if err != nil {
-				// rollback on error, keep the ClaimRef until we can successfully update the volume
-				volume.Spec.ClaimRef = oldClaimRef
+				return err
+			}
+			newVolume.Spec.ClaimRef = nil
+			if _, err = binderClient.UpdatePersistentVolume(newVolume); err != nil {
 				return fmt.Errorf("Unexpected error saving PersistentVolume: %+v", err)
 			}
+			volumeIndex.Update(newVolume)
 		}
 
-		_, exists, err := volumeIndex.Get(volume)
-		if err != nil {
-			return err
-		}
-		if !exists {
-			volumeIndex.Add(volume)
-		}
-		glog.V(5).Infof("PersistentVolume[%s] is now available\n", volume.Name)
 		nextPhase = api.VolumeAvailable
 
 	// available volumes await a claim
 	case api.VolumeAvailable:
 		// TODO:  remove api.VolumePending phase altogether
-		_, exists, err := volumeIndex.Get(volume)
-		if err != nil {
-			return err
-		}
-		if !exists {
-			volumeIndex.Add(volume)
-		}
 		if volume.Spec.ClaimRef != nil {
-			_, err := binderClient.GetPersistentVolumeClaim(volume.Spec.ClaimRef.Namespace, volume.Spec.ClaimRef.Name)
-			if err == nil {
-				// change of phase will trigger an update event with the newly bound volume
-				glog.V(5).Infof("PersistentVolume[%s] is now bound\n", volume.Name)
-				nextPhase = api.VolumeBound
-			} else {
-				if errors.IsNotFound(err) {
-					nextPhase = api.VolumeReleased
-				}
+			// an available volume should never have ClaimRef.  Stomp it so we end up in a consistent state.
```

Comment:
I think this is technically incorrect.

When a PV is bound, it's ClaimRef is set in the first transaction.  A second transaction will update its state to Bound, since persistent Spec and Status are different calls.

It is possible, therefore, to have an Available volume with a ClaimRef, at least for a very short while.